File: services/client/src/protocol/codec.go
package protocol

import "fmt"

/* 
 * Serializa un Header en una secuencia de bytes
 *
 * El header tiene un tamaño fijo de 5 bytes:
 *
 * El primer byte para el tipo de mensaje.
 * Los bytes 1 a 4 representan el tamaño del payload.
 */     
func EncodeHeader(header Header) ([]byte, error) {
	if err := ValidateHeader(header); err != nil {
		return nil, err
	}

	buffer := make([]byte, HeaderSize)

	// Obtengo el primer byte 
	buffer[0] = byte(header.Type)

	// Utilizo shifts para obtener cada uno de los 4 bytes
	buffer[1] = byte(header.PayloadLength >> 24)
	buffer[2] = byte(header.PayloadLength >> 16)
	buffer[3] = byte(header.PayloadLength >> 8)
	buffer[4] = byte(header.PayloadLength)

	return buffer, nil
}

/*
 * Deserializa una secuencia de bytes y reconstruye
 * el header correspondiente.
 */
func DecodeHeader(data []byte) (Header, error) {
	if len(data) != HeaderSize {
		return Header{}, fmt.Errorf(
			"invalid header size: expected %d, received %d",
			HeaderSize,
			len(data),
		)
	}

	// Obtengo el tipo de mensaje.
	messageType := MessageType(data[0])

	// Ahora obtengo los bytes del tamaño del payload
	//
	// Cada byte se convierte primero a uint32 y luego se desplaza
	// hasta la posición original
	payloadLength :=
		uint32(data[1])<<24 |
			uint32(data[2])<<16 |
			uint32(data[3])<<8 |
			uint32(data[4])

	header := Header{
		Type: messageType,
		PayloadLength: payloadLength,
	}

	if err := ValidateHeader(header); err != nil {
		return Header{}, err
	}

	return header, nil
}

File: services/server/src/protocol/codec.py
from .message import (
    HEADER_SIZE,
    MAX_PAYLOAD_SIZE,
    Header,
    MessageType,
)

"""Valida que el header respete las restricciones del protocolo"""
def validate_header(header: Header) -> None:
    if header.payload_length < 0:
        raise ValueError("payload length cannot be negative")

    if header.payload_length > MAX_PAYLOAD_SIZE:
        raise ValueError(
            f"payload too large: {header.payload_length} bytes"
        )


"""
Serializa un Header utilizando el formato del protocolo

Primer byte = tipo de mensaje
Siguientes 4 bytes = tamaño del payload
"""
def encode_header(header: Header) -> bytes:
    validate_header(header)

    message_type = bytes([header.message_type])

    payload_length = header.payload_length.to_bytes(
        4,
        byteorder="big", # Con esto me aseguro de que sea big endian
    )

    return message_type + payload_length


"""Reconstruye un Header a partir de sus 5 bytes"""
def decode_header(data: bytes) -> Header:
    if len(data) != HEADER_SIZE:
        raise ValueError(
            f"invalid header size: expected {HEADER_SIZE}, got {len(data)}"
        )

    # Obtengo el primer byte
    try:
        message_type = MessageType(data[0])
    except ValueError as error:
        raise ValueError(f"unknown message type: {data[0]}")

    # Obtengo el largo del payload
    payload_length = int.from_bytes(
        data[1:5],
        byteorder="big", # Con esto me aseguro de que sea big endian
    )

    header = Header(
        message_type=message_type,
        payload_length=payload_length,
    )

    validate_header(header)

    return header